torch/utils/data/dataset.py

```diff
@@ -253,6 +254,57 @@ def __len__(self):
         return total
 
 
+class ShuffleDataset(IterableDataset[T_co]):
+    r"""Dataset shuffled from the original dataset.
+
+    This class is useful to shuffle an existing instance of an IterableDataset.
+    The buffer with buffer_size is filled with the items from the dataset first. Then,
+    each item will be yielded from the buffer by reservoir sampling via iterator.
+
+    For buffer size smaller than 1, the buffer size is set to 1 representing
+    non-shuffled dataset. For perfect shuffling, the buffer size is required to be
+    greater than or equal to the size of dataset.
+
+    When it is used with :class:`~torch.utils.data.DataLoader`,each item in the
+    dataset will be yielded from the :class:`~torch.utils.data.DataLoader` iterator.
+
+    When it is used with :class:`~torch.utils.data.DataLoader` with :attr:`num_workers == 0`,
+    manually setting up a random seed for the single-process mode:
+
+        >>> ds = ShuffleDataset(dataset)
+        >>> random.seed(...)
+        >>> print(list(torch.utils.data.DataLoaderDataLoader(ds, num_workers=0)))
+
+    When it is used with :class:`~torch.utils.data.DataLoader` with :attr:`num_workers > 0`,
+    using :attr:`worker_init_fn` to set up a random seed for all workers:
+
+        >>> ds = ShuffleDataset(dataset)
+        >>> def init_fn(worker_id):
+        ...     random.seed(...)
+        >>> print(list(torch.utils.data.DataLoader(ds, ..., num_workers=n, worker_init_fn=init_fn)))
+
+    Arguments:
+        dataset (IterableDataset): The original IterableDataset.
+        buffer_size (int): The buffer size for shuffling.
+    """
+    dataset: IterableDataset[T_co]
+    buffer_size: int
+
+    def __init__(self, dataset: IterableDataset[T_co], buffer_size: int) -> None:
+        super(ShuffleDataset, self).__init__()
+        self.dataset = dataset
+        self.buffer_size = buffer_size if buffer_size > 1 else 1
+
+    def __iter__(self) -> Iterator[T_co]:
+        buf: List[T_co] = []
+        for x in self.dataset:
+            if len(buf) == self.buffer_size:
+                yield buf.pop(random.randint(0, self.buffer_size - 1))
+            buf.append(x)
+        while buf:
+            yield buf.pop(random.randint(0, len(buf) - 1))
```

Comment:
Nit: The same comment as above, performance-wise it would be better to shuffle list once and return it element by element.